### ISSUE TITLE
[BUG]: Docstring/signature of bound function not rendering properly

### ISSUE THREAD

**Issue by awni** [wjakob/nanobind#481]
### Problem description

The docstring does not render properly using built-in `help`.

### Reproducible example code

```text
In C++
++
#include <nanobind/nanobind.h>

namespace nb = nanobind;

int add(int x, int y) {
  return x + y;
}

NB_MODULE(test, m) {
    m.def("add", &add, "Add two numbers.");
}


```shell
>>> import test
>>> help(test.add)
```

Produces
```text
Help on nb_func in module test:

add = <nanobind.nb_func object>
    add(arg0: int, arg1: int, /) -> int
    
    Add two numbers.
```
```

**Comment by awni** [comment]
In addition to the strange looking doc with `help`, I notice that sphinx is picking up these as `autodata` instead of `autofunction` when using `autosummary`

**Comment by awni** [comment]
The thing that bothers me much more so than the `help` is it breaks our documentation because autosummary treats these with `autodata` instead of `autofunction`. I will see if I can find a workaround.

**Comment by wjakob** [comment]
I'm using nanobind to create bindings for a large and complex project called Dr.Jit, which uses Sphinxfor documentation. See https://drjit.readthedocs.io/en/nanobind_v2/reference.html for an example of the function reference and https://github.com/mitsuba-renderer/drjit/tree/nanobind_v2 for the GitHub repository (important: select `nanobind_v2 branch in each, `master` is the older pybind11-based version).

When it comes to Sphinx + nanobind, I really haven't encountered this problem you mention. Here is how the reference is generated, mostly with `auto*` and a few manually added docstrings: https://raw.githubusercontent.com/mitsuba-renderer/drjit/nanobind_v2/docs/reference.rst

**Comment by awni** [comment]
Yes well I looked at that project already and I think the difference is you use `autofunction` and we are using `autosummary` which attempts to infer the type of the python object to be documented.

**Comment by wjakob** [comment]
Aha. But that's an easy switch, no?

**Comment by awni** [comment]
This is the code that breaks: https://github.com/sphinx-doc/sphinx/blob/7894b5a2698c6abf4d285da78409785c40eab7b2/sphinx/ext/autodoc/__init__.py#L1304-L1310

**Comment by awni** [comment]
You can see it fails to infer that a `nanobind.nb_func` is actually a function since none of those conditions are `True`.

**Comment by awni** [comment]
It is also not so easy to switch because `autosummary` does some nice things like generate tables of functions with signatures: https://ml-explore.github.io/mlx/build/html/python/ops.html

We don't necessarily want to give that up and reimplementing it is some work I think.

**Comment by awni** [comment]
FYI this is the patch I came up with which seems to work if you put it in the Sphinx `conf.py` file:

```
def setup(app):
    wrapped = app.registry.documenters["function"].can_document_member
    def nanobind_function_patch(member: Any, membername: str, isattr: bool, parent: Any) -> bool:
        return "nanobind.nb_func" in str(type(member)) or wrapped(member)
    app.registry.documenters["function"].can_document_member = nanobind_function_patch
```

It's not pretty but it's concise enough.

**Comment by wjakob** [comment]
> You can see it fails to infer that a ``nanobind.nb_func`` is actually a function since none of those conditions are ``True``.

Exactly, the logic is essentially the same in `help()` (which is implemented in  `pydoc.py` that ships with the CPython release, and which also calls `inspect`). This also makes it clear how the logic cannot possibly be fixed from within the nanobind project since these are all CPython-specific types that are of no use in nanobind that needs a much deeper level of control of its function object. If it's something that really bothers you, then the thing to do is to escalate it to Sphinx, CPython (the `inspect.py` and/or `pydoc.py` maintainers). It will be more convincing coming from someone else and not me (I am obviously biased here).

> It is also not so easy to switch because autosummary does some nice things like generate tables of functions with signatures: https://ml-explore.github.io/mlx/build/html/python/ops.html

That is pretty cool, actually. I suppose the easiest thing to do as a workaround would be to fork Sphinx and change that line of code to get it to accept nanobind function objects as a function as well.

**Comment by awni** [comment]
> This also makes it clear how the logic cannot possibly be fixed from within the nanobind project

Yes I completely agree.. though you may want to document the change for people coming from Pybind11

> If it's something that really bothers you, then the thing to do is to escalate it to Sphinx, CPython (the inspect.py and/or pydoc.py maintainers). It will be more convincing coming from someone else and not me (I am obviously biased here).

Thanks! I will look into escalating. It seems like the right thing is to fix `inspect.isfunction` .. I don't have any idea how feasible it is for them but probably a good thing for them to be aware of.
